aspnetcore/blazor/security/passkeys-migration.md

```diff
@@ -0,0 +1,849 @@
+
+# [Visual Studio Code](#tab/visual-studio-code)
+
+Use the following command in the **Terminal** (**Terminal** menu > **New Terminal**) to add a migration for the new data annotations:
+
+```dotnetcli
+dotnet ef migrations add AddPasskeySupport
+```
+
+To apply the migration to the database, execute the following command:
+
+```dotnetcli
+dotnet ef database update
+```
+
+# [.NET CLI](#tab/net-cli/)
+
+To add a migration for the new data annotations, execute the following command in a command shell opened to the project's root folder:
+
+```dotnetcli
+dotnet ef migrations add AddPasskeySupport
+```
+
+To apply the migration to the database, execute the following command:
+
+```dotnetcli
+dotnet ef database update
+```
+
+---
+
+## Step 3: Create passkey model classes
+
+Add the following model classes to the project in the `Components/Account` folder.
+
+Replace the `{NAMESPACE}` placeholder in the following examples with the app's namespace. For example, `Contoso` for `Contoso.Components.Account` in the following examples.
+
+`Components/Account/PasskeyInputModel.cs`:
+
+```csharp
+namespace {NAMESPACE}.Components.Account;
+
+public class PasskeyInputModel
+{
+    public string? CredentialJson { get; set; }
+    public string? Error { get; set; }
+}
+```
+
+`Components/Account/PasskeyOperation.cs`:
+
+```csharp
+namespace {NAMESPACE}.Components.Account;
+
+public enum PasskeyOperation
+{
+    Create = 0,
+    Request = 1,
+}
+```
+
+## Step 4: Create the `PasskeySubmit` component
+
+Add the following `PasskeySubmit` component to handle passkey operations.
+
+`Components/Account/Shared/PasskeySubmit.razor`:
+
+```razor
+@using Microsoft.AspNetCore.Antiforgery
+@inject IServiceProvider Services
+
+<button type="submit" name="__passkeySubmit" 
+    @attributes="AdditionalAttributes">@ChildContent</button>
+<passkey-submit
+    operation="@Operation"
+    name="@Name"
+    email-name="@EmailName"
+    request-token-name="@tokens?.HeaderName"
+    request-token-value="@tokens?.RequestToken">
+</passkey-submit>
+
+@code {
+    private AntiforgeryTokenSet? tokens;
+
+    [CascadingParameter]
+    private HttpContext HttpContext { get; set; } = default!;
+
+    [Parameter]
+    [EditorRequired]
+    public PasskeyOperation Operation { get; set; }
+
+    [Parameter]
+    [EditorRequired]
+    public string Name { get; set; } = default!;
+
+    [Parameter]
+    public string? EmailName { get; set; }
+
+    [Parameter]
+    public RenderFragment? ChildContent { get; set; }
+
+    [Parameter(CaptureUnmatchedValues = true)]
+    public IDictionary<string, object>? AdditionalAttributes { get; set; }
+
+    protected override void OnInitialized()
+    {
+        tokens = Services.GetRequiredService<IAntiforgery>()?.GetTokens(HttpContext);
+    }
```

Comment:
This should be `GetService()` instead of `GetRequiredService()` (probably my fault).